packages/core/profiler/src/ApplicationProfiler.js

```diff
@@ -0,0 +1,82 @@
+// @flow strict-local
+
+import type {ApplicationProfilerEvent, IDisposable} from '@parcel/types';
+import type {
+  ApplicationProfilerMeasurement,
+  ApplicationProfilerMeasurementData,
+} from './types';
+import {ValueEmitter} from '@parcel/events';
+
+// $FlowFixMe
+import {performance as _performance} from 'perf_hooks';
+
+let tid;
+try {
+  tid = require('worker_threads').threadId;
+} catch {
+  tid = 0;
+}
+
+const performance: Performance = _performance;
+const pid = process.pid;
+
+export default class ApplicationProfiler {
+  #traceEmitter /* ValueEmitter<ApplicationProfilerEvent> */ =
+    new ValueEmitter();
+
+  #enabled /* boolean */ = false;
+
+  onTrace(cb: (event: ApplicationProfilerEvent) => mixed): IDisposable {
+    return this.#traceEmitter.addListener(cb);
+  }
+
+  async wrap(name: string, fn: () => mixed): Promise<void> {
+    let measurement = this.createMeasurement(name);
+    try {
+      await fn();
+    } finally {
+      measurement.end();
+    }
+  }
+
+  createMeasurement(
+    name: string,
+    data?: ApplicationProfilerMeasurementData = {categories: ['Core']},
+  ): ApplicationProfilerMeasurement {
+    const start = performance.now();
```

Comment:
Should this method just no-op if `this.enabled === false`? Alternatively, call sites could be gated by that same check, which isn't a bad idea given that it could avoid allocating the data objects if profiling is disabled.

Another idea (mentioned in my first comment) could be to add category-specific methods that take positional arguments, which could allow call sites to avoid creating the complex `data` object. Those methods could just no-op by default, or build the args and call `createMeasurement` otherwise.